proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,205 @@
+mode icons can and will be made automatically visible once the system setting is
+in dark mode, thereby enhancing the overall browser experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons, Firefox can specify icons).
+
+### Known Consumers
+
+The associated and linked bug has 48 stars and 31 comments.
+
+## Specification
+
+### Schema
+
+Some browsers don't currently have plans to support svg. Therefore, a drop-in
+replacement of an already supported image type will be allowed.
+
+numberToPathDictionary
+```
+{
+  "<number>": "<path>"
+}
+```
+
+manifest.json
+```
+{
+  "icon_variants": {
+    "<dark|light>": <numberToPathDictionary>
+  },
+  "action": {
+    "default_icon": <numberToPathDictionary>|<svgPathString|pathString>,
+    "icon_variants": {
+      "<dark|light>": <numberToPathDictionary>
+    }
+  }
+}
+```
+
+Action
+```
+action.setIcon({
+  imageData?: ImageData | object:<numberToPathDictionary>,
+  path?: string | object:<numberToPathDictionary>,
+  tabId?: number,
+  variants?: {
+    "<dark|light>": <numberToPathDictionary>
+  },
+});
+```
+
+### Examples
+
+manifest.json
+```
+{
+  "icon_variants": {
+    "dark": {
+      "128": "128_dark.png"
+    },
+    "light": {
+      "128": "128_light.png"
+    }
+  },
+  "action": {
+    "default_icon": {
+      "128": "128_action.png"
+    },
+    "icon_variants": {
+      "dark": {
+        "128": "128_action_dark.png"
+      },
+      "light": {
+        "128": "128_action_light.png"
+      }
+    }
+  }
+}
+```
+
+Action
+```
+action.setIcon({
+  variants?: {
+    "dark": {
+      "128": "128_action_dark.png"
+    },
+    "light": {
+      "128": "128_action_light.png"
+    }
+  },
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+#### Optional or required?
+`icon_variants` is an optional key of the top level manifest dictionary.
+`icon_variants` is also an optional sub-key in the action key dictionary.
+
+#### Fallback
+`icon_variants` will be used if they're supplied. Otherwise `default_icon` will
+be used if it's supplied. If neither are supplied, then `icons` will be used.
```

Comment:
About order of precedence and fallback.

1. If I understand correctly, the `icon_variants` or `variants` in `action.setIcon()` must provide both dark and light icon if they occur, right? This can simplify the fallback problem.

1. Order of precedence for action:
`API: action.setIcon()` → `action.icon_variants` → `action.default_icon` → `icon_variants` → `icons`

1. Order of precedence for non-action:
`icon_variants` → `icons`

Reply:
It would definitely be nice to have some more detail here. For example:

- What happens if you specify icon_variants.dark but not icon_variants.light, and also icons? Do we use the dark mode icon from icon_variants for light as well, or fallback to icons?
- What happens if you specify icon_variants.dark but not icon_variants.light, and you don't have icons? Do we use the dark mode icon from icon_variants or show an auto-generated icon?

Reply:
TBA (to be addressed) in upcoming patch. Can you clarify `Do we use the dark mode icon from icon_variants or show an auto-generated icon?`

Reply:
> Can you clarify Do we use the dark mode icon from icon_variants or show an auto-generated icon?

In this case you only explicitly specified a dark mode icon, so we can either use that for light mode as well, or just fallback to the browser's default behavior (which in Chrome is a grey icon with the first letter of the extension name).

Reply:
The new patch adds the following to one of your other related questions:

> Another option could be to fall back to
the dark mode icon if in light mode and no light mode icon is supplied.

I suppose that the proposal shouldn't necessarily have options in it, so it's an opportunity for input.